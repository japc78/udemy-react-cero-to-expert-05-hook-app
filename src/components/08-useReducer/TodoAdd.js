import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
	// Al manejar solo un campo (input) se puede hacer destructuracion del objeto y extraer la descripcion.
	const [{ description }, handelInputChange, reset ] = useForm({
		description: '',
	});
	// console.log(description);


	// Manejador del formulario.
	const handleSubmit = (e) => {
		e.preventDefault();

		if (description.trim().length <= 1) {
			return;
		}

		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false
		};

		handleAddTodo(newTodo);
		reset();
	}

	return (
		<>
			<h4>Agregar TODO</h4>
			<hr/>

			<form onSubmit= { handleSubmit }>
				<input
					type="text"
					name="description"
					onChange={ handelInputChange }
					className="form-control"
					placeholder="Aprender ..."
					value={ description }
					autoComplete="off"/>

				<button
					type="submit"
					className="btn btn-sm btn-outline-primary mt-1 btn-block"
				>
					Agregar
				</button>
			</form>
		</>
	)
}
