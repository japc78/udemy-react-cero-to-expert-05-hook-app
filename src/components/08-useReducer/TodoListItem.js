import React from 'react'

export const TodoListItem = ( {todo, idx, handleDelete, handleToggle }) => {
	return (
		// todoListItem, argumentos: todo, index, handleDelete, handleToogle
		<li
			key={todo.id}
			className="list-group-item"
		>
			<p
				onClick={ () => handleToggle(todo.id) }
				className={ `${ todo.done && 'complete'}` }>
					{idx + 1}. {todo.desc}
			</p>
			<button
				onClick= { () => handleDelete(todo.id) }
				className="btn btn-sm btn-danger">
					borrar
			</button>

		</li>
	)
}
