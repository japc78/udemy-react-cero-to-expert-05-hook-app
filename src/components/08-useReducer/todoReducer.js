

export const todoReducer = (state = [], action) => {
	switch (action.type) {
		case 'add':
			return [ ...state, action.payload]

		case 'delete':
			//filter retorna un nuevo array.
			// En este caso devuleve un array con todos aquelos elementos que sean distintos al id pasado por el payload.
			return state.filter(todo => todo.id !== action.payload);

		// Forma larga
		case 'toggle-old':
			return state.map(todo => {
				if (todo.id === action.payload) {
					return {
						...todo,
						done: !todo.done
					}

				}else {
					return todo;
				}
			})


		// Forma corta con return implicito en el map y ternario.
		case 'toggle':
			return state.map( todo =>
				( todo.id === action.payload )
					? {...todo, done: !todo.done}
					: todo
			)

		default:
			return state;
	}
}