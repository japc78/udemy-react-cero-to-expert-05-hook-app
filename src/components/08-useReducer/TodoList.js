import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ( { todos, handleDelete, handleToggle } ) => {
	return (
		/* TodoList, argumetnos: todos, handleDelete, handleToogle*/
		<ul className="list-group list-group-flush">
			{
				todos.map((todo, idx) => (
					<TodoListItem
						key = {todo.id}
						todo = { todo }
						idx = { idx }
						handleDelete = { handleDelete }
						handleToggle = { handleToggle }
					/>
				))
			}
		</ul>
	)
}
