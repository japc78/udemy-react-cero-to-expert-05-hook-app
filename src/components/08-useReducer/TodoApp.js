import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer'
import { useForm } from '../../hooks/useForm'

import { TodoList } from './TodoList'
import { TodoAdd } from './TodoAdd'

import './styles.css'
export const TodoApp = () => {

	// Estado inicial, se le pasa al useReducer
	// const initialState = [{
	// 	id: new Date().getTime(),
	// 	desc: 'Aprender React',
	// 	done: false
	// }]

	const init = () => {
		// return [{
		// 	id: new Date().getTime(),
		// 	desc: 'Aprender React',
		// 	done: false
		// }]

		// Se recoge la informacion guardada en el localstore
		// Si JSON.parse(localStorage.getItem('todos')) es null retornara un array vacio.
		return JSON.parse(localStorage.getItem('todos')) || []
	}

	const [ todos, dispatch ] = useReducer(todoReducer, [], init);
	// console.log(todos);




	// Se utiliza useEffect para que cada vez que se actualize el estado todos, se grabe en localstore.
	// Se pasa a JSON con stringify los todos ya que localstore solo permite guardar en String.
	useEffect( () => {
		localStorage.setItem('todos', JSON.stringify(todos))
	}, [todos])


	const handleDelete = ( todoId ) => {
		console.log("Aqui: " + todoId);

		//Tarea
		// Crear la action
		const action = {
			type: 'delete',
			payload: todoId
		}

		// hacer el dispach
		dispatch(action);
	}

	const handleToggle = (todoId) => {
		dispatch({
			type: 'toggle',
			payload: todoId
		})
	}


	const handleAddTodo = ( newTodo ) => {
		dispatch({
			type: 'add',
			payload: newTodo
		})
	}

	return (
		<div>
			<h1>TodoApp ({ todos.length })</h1>
			<hr/>

			<div className="row">
				<div className="col-7">
					<TodoList
						todos = { todos }
						handleDelete = { handleDelete }
						handleToggle = { handleToggle }
					/>
				</div>

				<div className="col-5">
					<TodoAdd handleAddTodo = { handleAddTodo }/>
				</div>
			</div>
		</div>
	)
}
