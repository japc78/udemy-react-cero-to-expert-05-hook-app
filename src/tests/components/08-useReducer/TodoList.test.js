import React from 'react'
import '@testing-library/jest-dom'
import { shallow } from 'enzyme'
import { TodoList } from '../../../components/08-useReducer/TodoList'
import { demoTodos } from './fixtures/demoTodos'

describe('Test on <TodoList/>', () => {

	const handleDelete = jest.fn();
	const handleToggle = jest.fn();

	const wrapper = shallow(
		<TodoList todos= { demoTodos }
			handleDelete = { handleDelete }
			handleToggle = { handleToggle }
		/>
	);


	test('should be shown correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});


	test('should have two <TodoListItem />', () => {
		// console.log(wrapper.find('TodoListItem').length);
		expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);

		// Se muestra por consola las props de uno de los componentes
		// console.log(wrapper.find('TodoListItem').at(0).props());

		// Se muestra una prop del componente
		// console.log(wrapper.find('TodoListItem').at(0).prop(handleDelete));

		// Comprueba que en la prop "handleDelete" del componente se espera una funcion.
		expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));
		// expect(wrapper.find('TodoListItem').at(1).prop('handleToggle')).toEqual(expect.any(Function));
	});
});
