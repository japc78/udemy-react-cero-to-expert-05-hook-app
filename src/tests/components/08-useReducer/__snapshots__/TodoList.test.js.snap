// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <TodoList/> should be shown correctly 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    idx={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    idx={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
